### PR TITLE
[Error] Produce a warning when users call external functions

### LINKED ISSUE #4001: Call python function in taichi scope raise TypeError: float() argument must be a string or a number, not 'Expr'
```python
import taichi as ti
ti.init(print_ir=True)

def vec3(*args):
    assert len(args) == 1 or len(args) == 3
    if len(args) == 1:
        x = float(args[0])
        result = ti.Vector([x, x, x])
    else:
        x, y, z = args
        result = ti.Vector([float(x), float(y), float(z)])

    return result


@ti.func
def process(a):
    x, y, z = a
    return vec3(x + 1, y + 1, z + 1)

@ti.kernel
def func():
    a = vec3(1, 2, 3)
    a = process(a)

func()
```

The above code raises Error

```
Traceback (most recent call last):
  File "test.py", line 26, in <module>
    func()
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/kernel_impl.py", line 715, in wrapped
    return primal(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/kernel_impl.py", line 643, in __call__
    key = self.ensure_compiled(*args)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/kernel_impl.py", line 629, in ensure_compiled
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/kernel_impl.py", line 475, in materialize
    taichi_kernel = _ti_core.create_kernel(taichi_ast_generator,
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/kernel_impl.py", line 465, in taichi_ast_generator
    transform_tree(tree, ctx)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/transform.py", line 8, in transform_tree
    ASTTransformer()(ctx, tree)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 26, in __call__
    raise e
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 23, in __call__
    return method(ctx, node)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer.py", line 484, in build_Module
    build_stmt(ctx, stmt)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 26, in __call__
    raise e
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 23, in __call__
    return method(ctx, node)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer.py", line 448, in build_FunctionDef
    build_stmts(ctx, node.body)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer.py", line 1083, in build_stmts
    build_stmt(ctx, stmt)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 26, in __call__
    raise e
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 23, in __call__
    return method(ctx, node)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer.py", line 69, in build_Assign
    build_stmt(ctx, node.value)
  File "/usr/local/lib/python3.8/dist-packages/taichi/lang/ast/ast_transformer_utils.py", line 32, in __call__
    raise TaichiCompilationError(msg)
taichi.lang.exception.TaichiCompilationError: On line 24 of file "test.py":
    a = process(a)
        ^^^^^^^^^^
On line 19 of file "test.py":
    return vec3(x + 1, y + 1, z + 1)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: float() argument must be a string or a number, not 'Expr'
```

### LINKED ISSUE #4008: Low python version for pylint job in CI
**Describe the bug**
The python version for the pylint job is 3.8. So, when I use a function that is introduced in 3.9, it produces the following error.

```
python/taichi/lang/ast/ast_transformer.py:371:31: E1101: Module 'ast' has no 'unparse' member (no-member)
```

### PR DESCRIPTION
**Pull request by lin-hitonami**
Related issue = close #4001, close #4008
For this program:
```python
import taichi as ti
ti.init()

def python_function():
    pass

@ti.kernel
def foo():
    python_function()

foo()
```
The warning is like this:
```
/home/lin/tmp/test.py:9: UserWarning: Calling non-taichi function "python_function". Scope inside the function is not be processed by the Taichi transformer. The function may not work as expected. Proceed with caution! Maybe you can consider turning it into a @ti.func?
  python_function()
```

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -242,6 +242,15 @@ def run(self):
                     os.remove(f)
 
 
+requirements = [
+    'numpy',
+    'sourceinspect>=0.0.4',
+    'colorama',
+    'astor',
+]
+if sys.version_info < (3, 9):
```
Comment (frostming) [inline setup.py:251]: Use an environment marker:
```python
requirements = [
    'numpy',
    'sourceinspect>=0.0.4',
    'colorama',
    'astor',
    'astunparse; python_version < "3.9"'
]
```
This is more friendly to dependency solvers.

File: python/taichi/lang/ast/ast_transformer.py [hunk 2]
```diff
@@ -329,34 +334,59 @@ def build_Call(ctx, node):
...
-            elif func_name == 'all':
-                node.ptr = ti_ops.ti_all(*args, **keywords)
-            else:
-                node.ptr = node.func.ptr(*args, **keywords)
+        func = node.func.ptr
+
+        if isinstance(node.func, ast.Attribute) and isinstance(
+                node.func.value.ptr, str) and node.func.attr == 'format':
+            args.insert(0, node.func.value.ptr)
+            node.ptr = impl.ti_format(*args, **keywords)
```
Comment (ailzhang) [inline python/taichi/lang/ast/ast_transformer.py:342]: Can we make this more organized by having a `build_builtin` or something similar? 
Main goal is to know where to find supported builtin functions in one place.

**Review by ailzhang — COMMENTED** [review]
Nice work!

File: python/taichi/lang/ast/ast_transformer_utils.py [hunk 3]
```diff
@@ -160,6 +180,9 @@ def is_in_static_for(self):
     def is_in_non_static(self):
         return self.non_static_status.is_in_non_static_scope
 
+    def is_in_static_scope(self):
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer_utils.py:183]: Why do we have both `StaticScopeGuard` and `NonStaticScopeGuard`?

File: python/taichi/lang/ast/ast_transformer.py [hunk 4]
```diff
@@ -329,34 +334,59 @@ def build_Call(ctx, node):
...
+                    file = None
+                if file and os.path.commonpath([
+                        file, package_root
+                ]) == package_root:  # functions inside taichi
+                    pass
+                else:
+                    name = unparse(node.func).strip()
+                    warnings.warn_explicit(
+                        f'Calling non-taichi function "{name}". '
+                        f'Scope inside the function is not be processed by the Taichi transformer. '
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:385]:
```suggestion
                        f'Scope inside the function is not processed by the Taichi AST transformer. '
```

File: python/taichi/lang/ast/ast_transformer_utils.py (hunk 3 above)
Comment (lin-hitonami) [reply python/taichi/lang/ast/ast_transformer_utils.py:183]: `StaticScopeGuard` is to track whether a statement is inside `ti.static`, and `NonStaticScopeGuard` is to track whether a statement is inside non-static `if`/`for`/`while`. Maybe we can rename `NonStaticScopeGuard` to `NonStaticControlFlowGuard` or something?

Comment (strongoier) [reply python/taichi/lang/ast/ast_transformer_utils.py:183]: Yep that will make things clearer.

**Review by ailzhang — APPROVED** [review]
Thanks!

**Pull request merged**